### PR TITLE
[Datasource, Resource] Network Address Groups, Member

### PR DESCRIPTION
**Pull request by michaelkad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
=== RUN   TestAccIBMPINetworkAddressGroupsDataSourceBasic
--- PASS: TestAccIBMPINetworkAddressGroupsDataSourceBasic (15.36s)
PASS

=== RUN   TestAccIBMPINetworkAddressGroupDataSourceBasic
--- PASS: TestAccIBMPINetworkAddressGroupDataSourceBasic (18.07s)
PASS

=== RUN   TestAccIBMPINetworkAddressGroupBasic
--- PASS: TestAccIBMPINetworkAddressGroupBasic (35.26s)
PASS

=== RUN   TestAccIBMPINetworkAddressGroupMemberBasic
--- PASS: TestAccIBMPINetworkAddressGroupMemberBasic (50.40s)
PASS
```

### CODE REVIEW

File: ibm/provider/provider.go [hunk 1]
```diff
@@ -649,6 +649,8 @@ func Provider() *schema.Provider {
 			"ibm_pi_instances":                              power.DataSourceIBMPIInstances(),
 			"ibm_pi_key":                                    power.DataSourceIBMPIKey(),
 			"ibm_pi_keys":                                   power.DataSourceIBMPIKeys(),
+			"ibm_pi_network_address_group":                  power.DataSourceIBMPINetworkAddressGroup(),
+			"ibm_pi_network_address_groups":                 power.DataSourceIBMPINetworkAddressGroups(),
 			"ibm_pi_network_port":                           power.DataSourceIBMPINetworkPort(),
```
Comment (ismirlia) [inline ibm/provider/provider.go:654]: I don't see the resource added to this list.

File: ibm/service/power/data_source_ibm_pi_network_address_group.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+			Arg_NetworkAddressGroupID: {
+				Description: "Network Address Group ID.",
+				Required:    true,
+				Type:        schema.TypeString,
+			},
+			// Attributes
+			Attr_CRN: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The Network Address Group's crn.",
```
Comment (ismirlia) [inline ibm/service/power/data_source_ibm_pi_network_address_group.go:39]: Fields here are unsorted.

File: ibm/service/power/data_source_ibm_pi_network_address_groups.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+			// Attributes
+			Attr_NetworkAddressGroups: {
+				Computed:    true,
+				Description: "list of Network Address Groups.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						Attr_CRN: {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The Network Address Group's crn.",
```
Comment (ismirlia) [inline ibm/service/power/data_source_ibm_pi_network_address_groups.go:39]: Fields here are unsorted.

File: website/docs/d/pi_network_address_group.html.markdown [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      zone      =   "lon04"
+    }
+  ```
+
+## Argument Reference
+
+You can specify the following arguments for this data source.
+
+- `pi_cloud_instance_id` - (Required, String) The GUID of the service instance associated with an account.
+- `pi_network_address_group_id` - (Required, String) network address group id or name.
```
Comment (ismirlia) [inline website/docs/d/pi_network_address_group.html.markdown:43]:
```suggestion
- `pi_network_address_group_id` - (Required, String) Network address group id or name.
```

File: website/docs/r/pi_network_address_group.html.markdown [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+- `pi_user_tags` - (Optional, List) A list of tags.
+
+## Attribute Reference
+
+In addition to all argument reference list, you can access the following attribute reference after your resource is created.
+
+- `crn` - (String) The network address group's crn.
+- `network_address_group_id` - (String) The unique identifier of the network address group.
+- `members` - (List) The list of IP addresses in CIDR notation in the network address group.
+Nested schema for `members`:
```
Comment (ismirlia) [inline website/docs/r/pi_network_address_group.html.markdown:55]:
```suggestion

  Nested schema for `members`:
```

File: website/docs/d/pi_network_address_groups.html.markdown [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+## Attribute Reference
+
+In addition to all argument reference list, you can access the following attribute references after your data source is created.
+
+- `network_address_groups` - (List) list of network address groups.
+Nested schema for **network_address_groups**:
+  - `crn` - (String) The network address group's crn.
+  - `id` - (String) The id of the network address group.
+  - `members` - (List) The list of IP addresses in CIDR notation (for example 192.168.66.2/32) in the network address group.
+ Nested schema for **members**:
```
Comment (ismirlia) [inline website/docs/d/pi_network_address_groups.html.markdown:52]:
```suggestion
 
   Nested schema for **members**:
```

File: website/docs/d/pi_network_address_groups.html.markdown [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+Review the argument references that you can specify for your data source.
+
+- `pi_cloud_instance_id` - (Required, String) The GUID of the service instance associated with an account.
+  
+## Attribute Reference
+
+In addition to all argument reference list, you can access the following attribute references after your data source is created.
+
+- `network_address_groups` - (List) list of network address groups.
+Nested schema for **network_address_groups**:
```
Comment (ismirlia) [inline website/docs/d/pi_network_address_groups.html.markdown:48]:
```suggestion

  Nested schema for **network_address_groups**:
```

File: website/docs/d/pi_network_address_group.html.markdown [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+## Attribute Reference
+
+In addition to all argument reference list, you can access the following attribute references after your data source is created.
+
+- `id` - The unique identifier of the pi_network_address_group.
+- `crn` - (String) The network address group's crn.
+
+- `members` - (List) The list of IP addresses in CIDR notation in the network address group.
+Nested schema for `members`:
```
Comment (ismirlia) [inline website/docs/d/pi_network_address_group.html.markdown:53]:
```suggestion

  Nested schema for `members`:
```

File: website/docs/d/pi_network_address_group_member.html.markdown [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+## Attribute Reference
+
+In addition to all argument reference list, you can access the following attribute reference after your resource is created.
+
+- `crn` - (String) The network address group's crn.
+- `network_address_group_id` - (String) The id of the network address group.
+- `members` - (List) The list of IP addresses in CIDR notation in the network address group.
+
+    Nested schema for `members`:
```
Comment (ismirlia) [inline website/docs/d/pi_network_address_group_member.html.markdown:56]:
```suggestion
  Nested schema for `members`:
```

File: ibm/provider/provider.go (hunk 1 above)
Comment (michaelkad) [reply ibm/provider/provider.go:654]: Good catch!

File: website/docs/d/pi_network_address_groups.html.markdown [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+Review the argument references that you can specify for your data source.
+
+- `pi_cloud_instance_id` - (Required, String) The GUID of the service instance associated with an account.
+  
+## Attribute Reference
+
+In addition to all argument reference list, you can access the following attribute references after your data source is created.
+
+- `network_address_groups` - (List) list of network address groups.
```
Comment (ismirlia) [inline website/docs/d/pi_network_address_groups.html.markdown:47]:
```suggestion
- `network_address_groups` - (List) List of network address groups.
```

**Review by ismirlia — APPROVED** [review]

File: ibm/service/power/resource_ibm_pi_network_address_group_member.go [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+				Type:         schema.TypeString,
+				ValidateFunc: validation.NoZeroValues,
+			},
+			Arg_Cidr: {
+				Description:  "The member to add in CIDR format.",
+				ExactlyOneOf: []string{Arg_Cidr, Arg_NetworkAddressGroupMemberID},
+				ForceNew:     true,
+				Optional:     true,
+				Type:         schema.TypeString,
+			},
```
Comment (Alexander-Kita) [inline ibm/service/power/resource_ibm_pi_network_address_group_member.go:47]: These should be switched to be in alphabetical order

File: website/docs/r/pi_network_address_group_member.html.markdown [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+- **create** - (Default 5 minutes) Used for creating network address group member.
+- **delete** - (Default 5 minutes) Used for deleting network address group member.
+
+## Argument Reference
+
+Review the argument references that you can specify for your resource.
+
+- `pi_cloud_instance_id` - (Required, String) The GUID of the service instance associated with an account.  
+- `pi_cidr` - (Optional, String) The member to add in CIDR format, for example 192.168.1.5/32. Required if `pi_network_address_group_member_id` not provided.
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_address_group_member.html.markdown:51]: Needs to be switched to be sorted

File: ibm/service/power/data_source_ibm_pi_network_address_group.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+	d.Set(Attr_CRN, networkAddressGroup.Crn)
+
+	members := []map[string]interface{}{}
+	if len(networkAddressGroup.Members) > 0 {
+		for _, mbr := range networkAddressGroup.Members {
+			member := memberToMap(mbr)
+			members = append(members, member)
+		}
+		d.Set(Attr_Members, members)
+	}
```
Comment (ismirlia) [inline ibm/service/power/data_source_ibm_pi_network_address_group.go:93]: did you not have to add to stop a double apply situation?
```
else  {
	d.Set(Attr_Members, nil)
}
```
I ran into that with rule-add code

Comment (michaelkad) [reply ibm/service/power/data_source_ibm_pi_network_address_group.go:93]: Updated

File: website/docs/r/pi_network_address_group_member.html.markdown (hunk 12 above)
Comment (michaelkad) [reply website/docs/r/pi_network_address_group_member.html.markdown]: Updated

File: ibm/service/power/resource_ibm_pi_network_address_group_member.go (hunk 11 above)
Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_network_address_group_member.go]: Updated

File: ibm/service/power/resource_ibm_pi_network_address_group.go [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+						Attr_ID: {
+							Computed:    true,
+							Description: "The id of the Network Address Group member IP addresses.",
+							Type:        schema.TypeString,
+						},
+					},
+				},
+				Optional: true,
+				Type:     schema.TypeList,
+			},
```
Comment (Alexander-Kita) [inline ibm/service/power/resource_ibm_pi_network_address_group.go:86]: These need to be switched

File: ibm/service/power/resource_ibm_pi_network_address_group_member_test.go [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+		Providers: acc.TestAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckIBMPINetworkAddressGroupMemberConfigBasic(name, cidr),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckIBMPINetworkAddressGroupMemberExists(nagMemberRes),
+					resource.TestCheckResourceAttrSet(nagMemberRes, power.Arg_NetworkAddressGroupID),
+					resource.TestCheckResourceAttr(nagMemberRes, power.Arg_Cidr, cidr),
+					resource.TestCheckResourceAttrSet(nagMemberRes, power.Attr_Name),
+				),
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_network_address_group_member_test.go:36]: Can you add a test for the member remove in the network-address-group?

File: ibm/service/power/ibm_pi_constants.go [hunk 16]
```diff
@@ -371,6 +377,7 @@ const (
 	Attr_UsedIPPercent                               = "used_ip_percent"
 	Attr_UsedMemory                                  = "used_memory"
 	Attr_UserIPAddress                               = "user_ip_address"
+	Attr_UserTags                                    = "user_tags"
```
Comment (ismirlia) [inline ibm/service/power/ibm_pi_constants.go:380]: I think this attr is no longer being used (at least in this PR).

File: ibm/service/power/resource_ibm_pi_network_address_group.go (hunk 14 above)
Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_network_address_group.go]: Done

File: ibm/service/power/ibm_pi_constants.go (hunk 16 above)
Comment (michaelkad) [reply ibm/service/power/ibm_pi_constants.go]: Removed.

File: ibm/service/power/resource_ibm_pi_network_address_group_member_test.go (hunk 15 above)
Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_network_address_group_member_test.go:36]: That is done in the destroy.

**Review by ismirlia — APPROVED** [review]

**Review by Alexander-Kita — APPROVED** [review]

File: ibm/service/power/resource_ibm_pi_network_address_group.go [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+func ResourceIBMPINetworkAddressGroup() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceIBMPINetworkAddressGroupCreate,
+		ReadContext:   resourceIBMPINetworkAddressGroupRead,
+		UpdateContext: resourceIBMPINetworkAddressGroupUpdate,
+		DeleteContext: resourceIBMPINetworkAddressGroupDelete,
+		Importer:      &schema.ResourceImporter{},
+
+		Timeouts: &schema.ResourceTimeout{
+			Delete: schema.DefaultTimeout(5 * time.Minute),
```
Comment (Alexander-Kita) [inline ibm/service/power/resource_ibm_pi_network_address_group.go:32]: Add create and update timeouts

File: website/docs/d/pi_network_address_group.html.markdown [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+In addition to all argument reference list, you can access the following attribute references after your data source is created.
+
+- `crn` - (String) The network address group's crn.
+- `id` - The unique identifier of the network address group.
+- `members` - (List) The list of IP addresses in CIDR notation in the network address group.
+
+  Nested schema for `members`:
+  - `cidr` - (String) The IP addresses in CIDR notation.
+  - `id` - (String) The id of the network address group member IP addresses.
+
```
Comment (Alexander-Kita) [inline website/docs/d/pi_network_address_group.html.markdown:56]:
```suggestion
```

File: website/docs/d/pi_network_address_groups.html.markdown [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+  Nested schema for `network_address_groups`:
+  - `crn` - (String) The network address group's crn.
+  - `id` - (String) The id of the network address group.
+  - `members` - (List) The list of IP addresses in CIDR notation in the network address group.
+
+      Nested schema for `members`:
+        - `cidr` - (String) The IP addresses in CIDR notation.
+        - `id` - (String) The id of the network address group member IP addresses.
+
```
Comment (Alexander-Kita) [inline website/docs/d/pi_network_address_groups.html.markdown:57]:
```suggestion
```

Comment (Alexander-Kita) [inline website/docs/d/pi_network_address_groups.html.markdown:57]:
```suggestion
```

File: website/docs/r/pi_network_address_group.html.markdown [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+
+- **delete** - (Default 5 minutes) Used for deleting network address group.
+
+## Argument Reference
+
+Review the argument references that you can specify for your resource.
+
+- `pi_cloud_instance_id` - (Required, String) The GUID of the service instance associated with an account.
+- `pi_name` - (Required, String) The name of the Network Address Group.
+- `pi_user_tags` - (Optional, List) A list of tags.
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_address_group.html.markdown:52]:
```suggestion
- `pi_user_tags` - (Optional, List) List of user tags attached to the resource.
```

File: website/docs/r/pi_network_address_group_member.html.markdown [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+
+In addition to all argument reference list, you can access the following attribute reference after your resource is created.
+
+- `crn` - (String) The network address group's crn.
+- `members` - (List) The list of IP addresses in CIDR notation in the network address group.
+
+    Nested schema for `members`:
+  - `cidr` - (String) The IP addresses in CIDR notation
+  - `id` - (String) The id of the network address group member IP addresses.
+
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_address_group_member.html.markdown:65]:
```suggestion
```

File: website/docs/r/pi_network_address_group.html.markdown [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+      region    =   "lon"
+      zone      =   "lon04"
+    }
+  ```
+
+## Timeouts
+
+ibm_pi_network_address_group provides the following [timeouts](https://www.terraform.io/docs/language/resources/syntax.html) configuration options:
+
+- **delete** - (Default 5 minutes) Used for deleting network address group.
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_address_group.html.markdown:44]: Add update and create timeout when added to resource.

File: ibm/service/power/resource_ibm_pi_network_address_group.go (hunk 17 above)
Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_network_address_group.go:32]: Not needed, here. The only one we are using is delete

File: website/docs/d/pi_network_address_groups.html.markdown (hunk 19 above)
Comment (michaelkad) [reply website/docs/d/pi_network_address_groups.html.markdown]: Done

File: website/docs/r/pi_network_address_group.html.markdown (hunk 20 above)
Comment (michaelkad) [reply website/docs/r/pi_network_address_group.html.markdown]: Done

File: website/docs/r/pi_network_address_group_member.html.markdown (hunk 21 above)
Comment (michaelkad) [reply website/docs/r/pi_network_address_group_member.html.markdown]: Done

File: website/docs/r/pi_network_address_group.html.markdown (hunk 22 above)
Comment (michaelkad) [reply website/docs/r/pi_network_address_group.html.markdown:44]: Not needed here

File: ibm/service/power/resource_ibm_pi_network_address_group.go [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+		hasChange = true
+	}
+	nagC := instance.NewIBMPINetworkAddressGroupClient(ctx, sess, parts[0])
+	if hasChange {
+		_, err := nagC.Update(parts[1], body)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_network_address_group.go:190]: Do we need a user tag update as well?

Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_network_address_group.go:190]: Yes!!

File: website/docs/d/pi_network_address_group.html.markdown [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+- `pi_cloud_instance_id` - (Required, String) The GUID of the service instance associated with an account.
+- `pi_network_address_group_id` - (Required, String) The network address group id.
+
+## Attribute Reference
+
+In addition to all argument reference list, you can access the following attribute references after your data source is created.
+
+- `crn` - (String) The network address group's crn.
+- `id` - The unique identifier of the network address group.
```
Comment (ismirlia) [inline website/docs/d/pi_network_address_group.html.markdown:50]:
```suggestion
- `id` - (String) The unique identifier of the network address group.
```

File: ibm/service/power/data_source_ibm_pi_network_address_groups.go [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+								},
+							},
+							Type: schema.TypeList,
+						},
+						Attr_Name: {
+							Computed:    true,
+							Description: "The name of the Network Address Group.",
+							Type:        schema.TypeString,
+						},
+					},
```
Comment (ismirlia) [inline ibm/service/power/data_source_ibm_pi_network_address_groups.go:73]: I think the Attr_UserTags itself is missing still.

Comment (michaelkad) [reply ibm/service/power/data_source_ibm_pi_network_address_groups.go:73]: Not needed here

**Comment by michaelkad** [comment]
```
=== RUN   TestAccIBMPINetworkAddressGroupBasic
--- PASS: TestAccIBMPINetworkAddressGroupBasic (38.50s)
PASS

=== RUN   TestAccIBMPINetworkAddressGroupMemberBasic
--- PASS: TestAccIBMPINetworkAddressGroupMemberBasic (52.32s)
PASS

=== RUN   TestAccIBMPINetworkAddressGroupDataSourceBasic
--- PASS: TestAccIBMPINetworkAddressGroupDataSourceBasic (18.30s)
PASS

=== RUN   TestAccIBMPINetworkAddressGroupsDataSourceBasic
--- PASS: TestAccIBMPINetworkAddressGroupsDataSourceBasic (16.49s)
PASS
```

**Review by ismirlia — DISMISSED** [review]

**Comment by michaelkad** [comment]
Opened in main repo

**Pull request closed without merging**